### ISSUE TITLE
4343081C - Resident Evil 4

### ISSUE THREAD

**Issue by Sowa95** [xenia-canary/game-compatibility#237]
### Validation

- [X] I've read the [FAQ](https://github.com/xenia-canary/xenia-canary/wiki/FAQ).
- [X] The Xenia build used is from a canary branch. (not MLBS/AlexVS/master/pull requests, etc)
- [X] This issue isn't for help (tech support) with Xenia, or a game.
- [X] I've checked if this game doesn't already have a compatibility report.

### Xbox 360 Marketplace link

https://marketplace.xbox.com/en-US/Product/Resident-Evil-4/66acd000-77fe-1000-9115-d8024343081c

### Xenia version

https://github.com/xenia-canary/xenia-canary/commit/4abbfa0b04a3739aa56eef8a5375cd9dea245011/

### Issues

- requires `ignore_thread_affinities = false` to load into gameplay.
- requires `delay_via_maybeyield = true` to improve performance after `ignore_thread_affinities`.
- press F5 or enable `clear_memory_page_state` to clear various artifacts.
- random crashes.
- game speed tied to FPS, don't disable `vsync`.

Biggest issue is stability, the game will die randomly in gameplay. Other than that, loading a save from gameplay will always result in a crash. Some cutscene skips also trigger a crash. On my PC it's impossible to get past the gate to the village. Also there are many other areas that are unable to load when I used saves from Nukernel. On the bright side - saving works, no issues with upscaling, Vulkan is on par with DX12.

### RE4 on Nukernel

I beat main campaign and Separate Ways on [Canary Nukernel](https://github.com/disjtqz/xenia-canar/releases), a highly experimental build that happens to work well with RE4. In about 35h of play time I got 2 crashes in gameplay, one in Chapter 2-3 just before boarding the cable lift with Ashley, and another one somewhere in the castle, but I don't remember where exactly. I also got a couple of crashes when I left the game AFK in equipment or pause menu for too long (about half an hour). It's almost as if Nukernel doesn't like to idle after controller turns itself off.

❗ `ignore_thread_affinities` and `delay_via_maybeyield` doesn't exist on Nukernel. It's not needed on this build.
❗ not recommended to touch `vsync` or `vsync_fps` on Nukernel, especially for RE4, as it will cause desync in cutscenes even if FPS is matching what the game wants.
❗ Nukernel isn't guaranteed to work well on every PC, don't take it for granted.

<details><summary>RE4 performance on Nukernel</summary>
<p>

Compared to main Canary, RE4 runs with terrible frame pacing on Nukernel. I suspect some part of the game is running too fast. Rivatuner limit at 30FPS will fix this issue - most of the game becomes smooth ([video](https://youtu.be/aurw0E3BxXM)). Unfortunately there are a bunch of areas heavily limited by CPU single thread performance, framerate tanks as low as 6-10FPS on 11900k. Worth to mention Vulkan is visibly faster than DX12, and considering both APIs are on par, it might be a better option.

I noticed 1 game breaking issue related to performance. In chapter 4-4 when Salazar statue starts chasing, QTE button mashing is very hard at low FPS. I couldn't beat it on Nukernel. For this section I switched to main Canary, then transferred the save back to Nukernel. I'm sure it's possible to complete this QTE on Nukernel if the FPS is bruteforced with better CPU. Experimenting with `time_scalar` might also help, but I haven't tested it.

</p>
</details>

<details><summary>A few softlocks.</summary>
<p>

**TLDR: If the game hangs trying to load specific cutscene, set 31FPS cap instead of 30FPS. In heavy areas framerate can struggle with 31FPS cap, I guess it's slightly too fast, but it's worth a shot to fix a stubborn cutscene. Switching between Vulkan/DX12 back and forth could also help.**

During main campaign I got a few harmless softlocks when transitioning from gameplay to cutscenes. When it happened, the game was still running, but was unable to load:
- cutscene in Chapter 4-1 after activating clock tower.
- cutscene in Chapter 5-2 after jumping with Ashley into trash disposal.
- cutscene after beating Sadler.

All of the above can be skipped with Back button, which resumes the game, so I didn't think much of it at first. But when I began Separate Ways I got a softlock that was impossible to skip. In Chapter 1 after completing RGB puzzle in the church, game would consisntently hang. Eventually I managed to go past it on Vulkan. I thougth maybe this is the fix, but later in Chapter 5 right before Leon strangles Ada passionately, the game refused to load FMV again. This time I couldn't load that particular cutscene in 6 attempts, switching between DX12 and Vulkan did not help. But then I capped the game at 31FPS instead of 30FPS, and it went through. 31FPS lock will cause cutscenes to progressively get out of sync ever so slightly, causing the cutscene to stop playing in last second or two (depending on lenght), but for some reason it worked. Wether it's luck or actually helps synchronize start of the cutscene, I don't know. I completed whole SW Chapter 5 with 31FPS limit.

</p>
</details>

Save with completed main campaign and Separate Ways:
[RE4.zip](https://github.com/user-attachments/files/16739557/RE4.zip)

### Log

[xenia.zip](https://github.com/user-attachments/files/16739545/xenia.zip)

### Screenshots

**All screenshots taken on Nukernel.**
![1](https://github.com/user-attachments/assets/db20742d-afa9-40ff-9d16-db453f78532f)
<details><summary>More screenshots</summary>
<p>

![2](https://github.com/user-attachments/assets/1dbb96e5-517f-4353-a719-aaaa3d94c4a4)
![3](https://github.com/user-attachments/assets/06468bc0-002d-42a2-b49e-75696b1d90b1)
![4](https://github.com/user-attachments/assets/bb297564-900a-49e4-be65-7e88999abe67)
![5](https://github.com/user-attachments/assets/afba7289-68eb-451a-85f3-1e714074ce8d)
![6](https://github.com/user-attachments/assets/20b09079-3c96-445b-a69b-0037b2aa11d8)

**DX12 vs Vulkan:**
![dxvk](https://github.com/user-attachments/assets/8dff341e-2e8a-4c3d-b918-db8e2abdff4e)

</p>
</details> 

### Labels

state-gameplay, tech-engine-mt, vsync-off-speedup, tech-middleware-criware

**Comment by BorealisEcho** [comment]
User Marmatica on Discord reports that the game is playable with settings mentioned by Sowa95 above.

There are texture issues and z-fighting in the final boss area. Fixable by using `render_target_path="ROV"`.

Occasional crashes have been noted (2x in the playthrough).

![image.png](https://github.com/user-attachments/assets/5db8133d-2397-43f3-802c-b25ec4f28119)

<img width="1275" height="768" alt="Image" src="https://github.com/user-attachments/assets/b1e34873-aea9-4f61-810c-6e68e28c3ec4" />
